File: S03/week06/sejun/maximumSubArray.java
// Time Limit Exceeded..

class Solution {
		public int maxSubArray(int[] nums) {
			int[][] sum = new int[2][100001];
			int size = nums.length;
			int max = Integer.MIN_VALUE;
			for (int i = 0; i < size; i++) {
				sum[0][i] = nums[i];
				max = max >= sum[0][i] ? max : sum[0][i];
			}
			for (int i = 1; i < size; i++) {
				for (int j = 0; j < size - i; j++) {
					sum[i%2][j] = sum[(i-1)%2][j] + nums[i+j];
					max = max >= sum[i%2][j] ? max : sum[i%2][j];
				}
			}

			return max;
		}
	}

File: S03/week06/z0/MaximumSubarray.kt
class Solution {
    // 413ms(87.37%)
    // 54.35MB(60.38%)
    fun maxSubArray(nums: IntArray): Int {
        var totalMax = Int.MIN_VALUE
        var subArrayMax = 0
        nums.forEach {
            // n번째까지의 (최대)합과 n번째 값을 비교
            // M(n) = Max(M(n-1)+n,n)
            subArrayMax = max(subArrayMax + it, it)
            // Max(M(0),..,M(n))
            totalMax = max(subArrayMax, totalMax)
        }
        return totalMax
    }
}
